File: Daily Challanges/2026-03-March/2026-03-27.py
def truncate_text(s):
    def char_width(c):
        if c in "ilI":
            return 1
        elif c in "fjrt":
            return 2
        elif c in "abcdeghkmnopqrstuvwxyzJL":
            return 3
        elif c in "ABCDEFGHKMNOPQRSTUVWXYZ":
            return 4
        elif c == " ":
            return 2
        elif c == ".":
            return 1
        return 0

    total = sum(char_width(c) for c in s)
    if total <= 50:
        return s

    max_text_width = 47  # 50 - 3 (for "...")
    running = 0
    cutoff = 0
    for i, c in enumerate(s):
        w = char_width(c)
        if running + w > max_text_width:
            break
        running += w
        cutoff = i + 1
    return s[:cutoff] + "..."

print(truncate_text("The quick brown fox"))
print(truncate_text("The fast striped zebra"))

File: Daily Challanges/2026-03-March/2026-03-27.js
function truncateText(s) {
  function charWidth(c) {
    if ("ilI".includes(c)) return 1;
    else if ("fjrt".includes(c)) return 2;
    else if ("abcdeghkmnopqrstuvwxyzJL".includes(c)) return 3;
    else if ("ABCDEFGHKMNOPQRSTUVWXYZ".includes(c)) return 4;
    else if (c === " ") return 2;
    else if (c === ".") return 1;
    return 0;
  }

  const total = [...s].reduce((sum, c) => sum + charWidth(c), 0);
  if (total <= 50) {
    return s;
  }

  const maxTextWidth = 47; // 50 - 3 (for "...")
  let running = 0;
  let cutoff = 0;
  for (let i = 0; i < s.length; i++) {
    const w = charWidth(s[i]);
    if (running + w > maxTextWidth) break;
    running += w;
    cutoff = i + 1;
  }

  return s.slice(0, cutoff) + "...";
}

console.log(truncateText("The quick brown fox"));
console.log(truncateText("The fast striped zebra"));
